README.md
# binary_codec
Dart library for converting standard data types to a binary format and back

Supported datatypes: Null, bool, double, int, Uint8List, String, Map, List. Maps and Lists may contain any of the supported datatypes as keys and values.

On dart2js ints are always encoded as doubles. And if you try to decode ints on dart2js out of the safe range you will probably not get the exact value. But as long as you stay on one platform (not switching between dartvm and dart2js) your on the safe side and don't need to think about it.


CHANGELOG.md
# 1.0.X

Not safe, don't use them.

# 2.0.1

Forgot the CHANGELOG.md :)
Now I've rewritten all the code, so it makes more sense, and does more of the things we expect it to do.
Changed the license to MIT since there are some hours invested in this code know.

# 2.0.2

run formatter


lib/src/byte_writer.dart
import 'dart:typed_data';

class ByteWriter {
  static const _initialLength = 8;

  var _buffer = Uint8List(_initialLength);
  var _length = _initialLength;
  var _index = 0;

  int get currentLength => _index;

  void _reserve(int byteCount) {
    assert(byteCount != null);
    assert(byteCount >= 0);
    while (_index + byteCount > _length) {
      _length *= 2;
    }
    if (_length != _buffer.length) {
      var next = Uint8List(_length);
      next.setRange(0, _index, _buffer);
      _buffer = next;
    }
  }

  void write(int byte) {
    assert(byte != null);
    _reserve(1);
    _buffer[_index++] = byte;
  }

  void writeAll(Iterable<int> bytes) {
    assert(bytes != null);
    _reserve(bytes.length);
    _buffer.setRange(_index, _index + bytes.length, bytes);
    _index += bytes.length;
  }

  Uint8List done() {
    var result = Uint8List.view(_buffer.buffer, 0, _index);
    _buffer = Uint8List(_initialLength);
    _index = 0;
    return result;
  }
}


lib/src/binary_codec.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'byte_reader.dart';
import 'byte_writer.dart';
import 'tags.dart';

class BinaryCodec extends Codec<dynamic, Uint8List> {
  const BinaryCodec();

  @override
  Converter<Uint8List, dynamic> get decoder => Decoder();

  @override
  Converter<dynamic, Uint8List> get encoder => Encoder();
}

class Decoder extends Converter<Uint8List, dynamic> {
  @override
  dynamic convert(Uint8List input) {
    return decode(ByteReader(input));
  }

  dynamic decode(ByteReader reader) {
    int tag = reader.read();
    if (tag < Tags.vNegInt) {
      return tag - Tags.vPosInt;
    } else if (tag < Tags.vShortBlob) {
      return -(tag - Tags.vNegInt);
    } else if (tag < Tags.vShortText) {
      return decodeBlob(tag - Tags.vShortBlob, reader);
    } else if (tag < Tags.vInt) {
      return decodeText(tag - Tags.vShortText, reader);
    } else if (tag < Tags.vNull) {
      return decodeInt(tag, reader);
    } else if (tag == Tags.vNull) {
      return null;
    } else if (tag == Tags.vTrue) {
      return true;
    } else if (tag == Tags.vFalse) {
      return false;
    } else if (tag == Tags.vDouble) {
      return decodeDouble(reader);
    } else if (tag == Tags.vBlob) {
      return decodeBlob(decodeLength(reader), reader);
    } else if (tag == Tags.vText) {
      return decodeText(decodeLength(reader), reader);
    } else if (tag == Tags.vList) {
      return decodeList(reader);
    } else if (tag == Tags.vMap) {
      return decodeMap(reader);
    } else {
      throw 'Tag \'$tag\' not handled';
    }
  }

  int decodeInt(int tag, ByteReader reader) {
    int byteCountMinusOne = tag - Tags.vInt;
    var value = 0;
    for (int i = 0; i < byteCountMinusOne; i++) {
      value += pow(2, i * 8) * reader.read();
    }
    var last = reader.read();
    value += pow(2, byteCountMinusOne * 8) * (last & 0x7F);
    if (last & 0x80 != 0) {
      value -= pow(2, byteCountMinusOne * 8 + 7);
    }
    return value;
  }

  double decodeDouble(ByteReader reader) {
    var bytes = reader.readAll(8);
    return bytes.buffer
        .asByteData(bytes.offsetInBytes, 8)
        .getFloat64(0, Endian.little);
  }

  int decodeLength(ByteReader reader) {
    int tag = reader.read();
    if (tag < Tags.vNegInt) {
      return tag - Tags.vPosInt;
    } else if (tag < Tags.vInt) {
      throw 'Tag \'$tag\' is no length';
    } else if (tag < Tags.vNull) {
      return decodeInt(tag, reader);
    } else if (tag == Tags.vDouble) {
      // Also expect doubles as length for dart2js
      var length = decodeDouble(reader);
      var rounded = length.round();
      if (length != rounded) {
        throw 'Tag \'$tag\' is a double value ($length) and no length (lengths must have no decimal places)';
      }
      return rounded;
    } else {
      throw 'Tag \'$tag\' is no length';
    }
  }

  Uint8List decodeBlob(int length, ByteReader reader) {
    return reader.readAll(length);
  }

  String decodeText(int length, ByteReader reader) {
    return utf8.decode(decodeBlob(length, reader));
  }

  List decodeList(ByteReader reader) {
    int length = decodeLength(reader);
    var list = List();
    for (int i = 0; i < length; i++) {
      list.add(decode(reader));
    }
    return list;
  }

  Map decodeMap(ByteReader reader) {
    int length = decodeLength(reader);
    var map = Map();
    for (int i = 0; i < length; i++) {
      var key = decode(reader);
      var value = decode(reader);
      map[key] = value;
    }
    return map;
  }
}

class Encoder extends Converter<dynamic, Uint8List> {
  @override
  Uint8List convert(dynamic value) {
    ByteWriter writer = ByteWriter();
    encode(value, writer);
    return writer.done();
  }

  void encode(dynamic value, ByteWriter writer) {
    if (value == null) {
      writer.write(Tags.vNull);
    } else if (value is bool) {
      writer.write(value ? Tags.vTrue : Tags.vFalse);
    } else if (value is double) {
      // Check for doubles first because of dart2js
      encodeDouble(value, writer);
    } else if (value is int) {
      encodeInt(value, writer);
    } else if (value is Uint8List) {
      encodeBlob(value, writer);
    } else if (value is String) {
      encodeText(value, writer);
    } else if (value is List) {
      encodeList(value, writer);
    } else if (value is Map) {
      encodeMap(value, writer);
    } else {
      throw 'Type of $value is not supported (${value.runtimeType})';
    }
  }

  void encodeDouble(double value, ByteWriter writer) {
    writer.write(Tags.vDouble);
    var bytes = Uint8List(8);
    bytes.buffer.asByteData().setFloat64(0, value, Endian.little);
    writer.writeAll(bytes);
  }

  void encodeInt(int value, ByteWriter writer) {
    if (value >= 0 && value < 128) {
      writer.write(Tags.vPosInt + value);
    } else if (value < 0 && value > -64) {
      writer.write(Tags.vNegInt - value);
    } else {
      int byteCountMinusOne = (value.bitLength / 8).floor();
      writer.write(Tags.vInt + byteCountMinusOne);
      bool neg = value < 0;
      if (neg) {
        value = -value - 1;
      }
      for (int i = 0; i <= byteCountMinusOne; i++) {
        var byte = value ~/ pow(2, i * 8);
        if (neg) {
          byte = -(byte + 1);
        }
        writer.write(byte);
      }
    }
  }

  void encodeLength(int length, ByteWriter writer) {
    // Store lengths in dart2js as doubles
    if (identical(0, 0.0)) {
      encodeDouble(length.toDouble(), writer);
    } else {
      encodeInt(length, writer);
    }
  }

  void encodeBlob(Uint8List value, ByteWriter writer) {
    encodeData(value, false, writer);
  }

  void encodeText(String value, ByteWriter writer) {
    encodeData(utf8.encode(value), true, writer);
  }

  void encodeData(Uint8List value, bool isText, ByteWriter writer) {
    var length = value.length;
    if (length < 24) {
      int tag = isText ? Tags.vShortText : Tags.vShortBlob;
      writer.write(tag + length);
    } else {
      writer.write(isText ? Tags.vText : Tags.vBlob);
      encodeLength(length, writer);
    }
    writer.writeAll(value);
  }

  void encodeList(List value, ByteWriter writer) {
    writer.write(Tags.vList);
    encodeLength(value.length, writer);
    for (var element in value.toList()) {
      encode(element, writer);
    }
  }

  void encodeMap(Map value, ByteWriter writer) {
    writer.write(Tags.vMap);
    encodeLength(value.length, writer);
    for (var entry in value.entries.toList()) {
      encode(entry.key, writer);
      encode(entry.value, writer);
    }
  }
}


example/main.dart
import 'dart:math';
import 'dart:typed_data';

import '../lib/binary_codec.dart';

/// Example text
const loremIpsum = 'Lorem ipsum dolor sit amet, consetetur sadipscing elitr, '
    'sed diam nonumy eirmod tempor invidunt ut labore et dolore '
    'magna aliquyam erat, sed diam voluptua. At vero eos et accusam '
    'et justo duo dolores et ea rebum. Stet clita kasd gubergren, no '
    'sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem '
    'ipsum dolor sit amet, consetetur sadipscing elitr, sed diam '
    'nonumy eirmod tempor invidunt ut labore et dolore magna '
    'aliquyam erat, sed diam voluptua. At vero eos et accusam '
    'et justo duo dolores et ea rebum. Stet clita kasd gubergren, '
    'no sea takimata sanctus est Lorem ipsum dolor sit amet.';

void main() {
  var exampleObject = generateSomeData();

  /// Encode the generated data
  Uint8List encoded = binaryCodec.encode(exampleObject);

  /// Decode the encoded data
  var decoded = binaryCodec.decode(encoded);

  /// Check if they are the same
  if (exampleObject.toString() == decoded.toString()) {
    /// They should be the same
    print('Success');
  } else {
    /// Else theres a bug in the library
    /// Please send the [exampleObject] that leads here and on which 
    /// platform you are to the author of this library (stefan.zemljic@gmx.ch)
    print('Wait... What???');
  }
}

dynamic generateSomeData() {
  var random = Random();
  var shortBinaryLists = List.generate(16, (length) {
    return Uint8List.fromList(
      List.generate(length, (_) => random.nextInt(256)),
    );
  });
  var someDoubles = [
    2930543.324960,
    -134943.2345,
    0.0,
    1.0,
    -1.0,
    double.infinity,
    double.nan,
    double.maxFinite,
    double.negativeInfinity,
    double.minPositive,
  ];
  var someInts = [321, 43956, -23459, 0, 1, -1, -239458];
  return {
    'small positive ints': List.generate(128, (x) => x),
    'small negative ints': List.generate(64, (x) => -x),
    'short binary lists': shortBinaryLists,
    'short texts': ['hello world', 'lorem ipsum', 'asdf qwer jklö'],
    'ints': someInts,
    'null': null,
    'booleans': [true, false],
    'doubles': someDoubles,
    'bigger binary lists':
        Uint8List.fromList(List.generate(1266, (x) => x % 256)),
    'bigger texts': loremIpsum.substring(0, 430),
    'lists': ['1', 2, 3.0, List.generate(4, (_) => 4), 'Five'],
    'maps': {'key1': 'value1', true: false, false: true, 4.0: 4, 5: 5.0},
  };
}
